### PR TITLE
Another attempt at an astable flag

### PR DESCRIPTION
**Pull request by pdeffebach**
This attempt should be a more robust strategy than the current `@t` flag in DataFrameMacros.jl. 

I will outline it in more detail shortly.

### CODE REVIEW

**Comment by pdeffebach** [comment]
More progress! Not ready for a review yet as i have not added a complete test set or begun documenting.

**Comment by pdeffebach** [comment]
Ready for a review!

This is a big PR. In the process, I have to finally remove some deprecated functionality for grouped data frames. In particular

```
@by df :a (n1 = :a, n2 = :b)
```

used to work. In the past release we had a warning that this needed to be 

```
@by df :a $AsTable = (n1 = :a, n2 = :b)
```

and now we throw an error in the original expression. 

Here is the docstring for `@astable`

---

## `astable(args...)`

Return a `NamedTuple` from a transformation inside DataFramesMeta.jl macros.

`@astable` acts on a single block. It works through all top-level expressions
and collects all such expressions of the form `:y = x`, i.e. assignments to a
`Symbol`, which is a syntax error outside of the macro. At the end of the
expression, all assignments are collected into a `NamedTuple` to be used
with the `AsTable` destination in the DataFrames.jl transformation
mini-language.

Concretely, the expressions

```
df = DataFrame(a = 1)

@rtransform df @astable begin
    :x = 1
    y = 50
    :z = :x + y + :a
end
```

becomes the pair

```
function f(a)
    x_t = 1
    y = 50
    z_t = x_t + y + a

    (; x = x_t, z = z_t)
end

transform(df, [:a] => f => AsTable)
```

`@astable` is useful when performing intermediate calculations
yet store their results in new columns. For example, the following fails.

```
@rtransform df begin
    :new_col_1 = :x + :y
    :new_col_2 = :new_col_1 + :z
end
```

This because DataFrames.jl does not guarantee sequential evaluation of
transformations. `@astable` solves this problem

@rtransform df @astable begin
    :new_col_1 = :x + :y
    :new_col_2 = :new_col_1 + :z
end

### Examples

```
julia> df = DataFrame(a = [1, 2, 3], b = [4, 5, 6]);

julia> d = @rtransform df @astable begin
           :x = 1
           y = 5
           :z = :x + y
       end
3×4 DataFrame
 Row │ a      b      x      z
     │ Int64  Int64  Int64  Int64
─────┼────────────────────────────
   1 │     1      4      1      6
   2 │     2      5      1      6
   3 │     3      6      1      6

julia> df = DataFrame(a = [1, 1, 2, 2], b = [5, 6, 70, 80]);

julia> @by df :a @astable begin
           $(DOLLAR)"Mean of b" = mean(:b)
           $(DOLLAR)"Standard deviation of b" = std(:b)
       end
2×3 DataFrame
 Row │ a      Mean of b  Standard deviation of b
     │ Int64  Float64    Float64
─────┼───────────────────────────────────────────
   1 │     1        5.5                 0.707107
   2 │     2       75.0                 7.07107
```

---

This implementation is more complicated than that of `@t` from DataFrameMacros.jl. In DataFrameMacros.jl, the following will fail

```
df = DataFrame(a = 1)
@transform df @t begin 
    :x = 1
    b + :x 
end
```

this is because it sends `:x` to `src` in `src => fun => dest` even though it doesn't exist in the DataFrame. `@astable` does not have this problem, at the cost of a more complicated implementation. 

cc @bkamins @nalimilan

**Comment by bkamins** [comment]
> `transform(df, [:a] => f => AsTable)`

I would think it should be `transform(df, [:a] => ByRow(f) => AsTable)`?

**Comment by pdeffebach** [comment]
Good catch. Will update.

File: src/macros.jl [hunk 1]
```diff
@@ -350,6 +350,99 @@ macro passmissing(args...)
...
+   1 │     1      4      1      6
+   2 │     2      5      1      6
+   3 │     3      6      1      6
+
+julia> df = DataFrame(a = [1, 1, 2, 2], b = [5, 6, 70, 80]);
+
+julia> @by df :a @astable begin
+           $(DOLLAR)"Mean of b" = mean(:b)
+           $(DOLLAR)"Standard deviation of b" = std(:b)
+       end
```
Comment (nalimilan) [inline src/macros.jl:432]: This example can be achieved without `@astable`, right? Maybe do `m = mean(:b); std(:b, mean=m)` to illustrate  the power of this function? Or, simpler, call `extrema(:b)` to create two columns.

Also, I wouldn't use long column names with spaces in them: better illustrate a single feature at a time.

File: src/parsing_astable.jl [hunk 2]
```diff
@@ -0,0 +1,95 @@
+function conditionally_add_symbols!(inputs_to_function, lhs_assignments, col)
```
Comment (nalimilan) [inline src/parsing_astable.jl:1]: How about making the signature more explicit? At least, `lhs_assignments::AbstractDict` would make things clearer. If you rely on the order, better even use `::OrderedCollections.OrderedDict`. Same below.

File: src/parsing_astable.jl [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    # if it's already been assigned at top-level,
+    # don't add it to the inputs
+    if haskey(lhs_assignments, col)
+        return lhs_assignments[col]
+    else
+        return addkey!(inputs_to_function, col)
+    end
```
Comment (nalimilan) [inline src/parsing_astable.jl:8]:
```suggestion
    return get!(lhs_assignments, col) do
        gensym()
    end
```

File: src/parsing_astable.jl [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    exprs = map(ex.args) do arg
+        if is_column_assigment(arg)
+            lhs = get_column_expr(arg.args[1])
+            rhs = arg.args[2]
+            new_ex = replace_syms_astable!(inputs_to_function, lhs_assignments, arg.args[2])
+            if haskey(inputs_to_function, lhs)
+                new_lhs = inputs_to_function[lhs]
+                lhs_assignments[lhs] = new_lhs
+            else
+                new_lhs = addkey!(lhs_assignments, lhs)
```
Comment (nalimilan) [inline src/parsing_astable.jl:65]: `addkey!` checks whether `lhs_assignments` contains `lhs`. If you already know that's not the case, better assign `gensym()` directly.

File: src/parsing.jl [hunk 5]
```diff
@@ -269,7 +270,13 @@ function fun_to_vec(ex::Expr;
...
     end
 
-    if no_dest
+    if final_flags[ASTABLE_SYM][]
+        src, fun = get_source_fun_astable(ex; exprflags = final_flags)
+
+        return :($src => $fun => AsTable)
+    end
+
+    if no_dest # subet and with
```
Comment (nalimilan) [inline src/parsing.jl:279]:
```suggestion
    if no_dest # subset and with
```

File: test/astable_flag.jl [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+
+const ≅ = isequal
+
+@testset "@astable with just assignments" begin
+    df = DataFrame(a = 1, b = 2)
+
+    d = @rtransform df @astable begin
+        :x = 1
+
+        nothing
```
Comment (nalimilan) [inline test/astable_flag.jl:15]: Why return `nothing`?

File: src/macros.jl [hunk 7]
```diff
@@ -350,6 +350,99 @@ macro passmissing(args...)
     throw(ArgumentError("@passmissing only works inside DataFramesMeta macros."))
 end
 
+"""
+    astable(args...)
+
+Return a `NamedTuple` from a transformation inside DataFramesMeta.jl macros.
+
+`@astable` acts on a single block. It works through all top-level expressions
+and collects all such expressions of the form `:y = x`, i.e. assignments to a
```
Comment (nalimilan) [inline src/macros.jl:359]:
```suggestion
and collects all such expressions of the form `:y = ...`, i.e. assignments to a
```

File: src/macros.jl [hunk 8]
```diff
@@ -350,6 +350,99 @@ macro passmissing(args...)
...
+df = DataFrame(a = 1)
+
+@rtransform df @astable begin
+    :x = 1
+    y = 50
+    :z = :x + y + :a
+end
+```
+
+becomes the pair
```
Comment (nalimilan) [inline src/macros.jl:377]:
```suggestion
become the pair
```

File: src/macros.jl [hunk 9]
```diff
@@ -350,6 +350,99 @@ macro passmissing(args...)
...
+
+```
+@rtransform df begin
+    :new_col_1 = :x + :y
+    :new_col_2 = :new_col_1 + :z
+end
+```
+
+This because DataFrames.jl does not guarantee sequential evaluation of
+transformations. `@astable` solves this problem
```
Comment (nalimilan) [inline src/macros.jl:402]: While this is an interesting side-effect, the main goal of `AsTable` is to allow returning multiple columns from a single "function". Probably worth mentioning? For example it's useful with `extrema` to compute the minimum and the maximum at the same time.

File: src/macros.jl [hunk 10]
```diff
@@ -350,6 +350,99 @@ macro passmissing(args...)
     throw(ArgumentError("@passmissing only works inside DataFramesMeta macros."))
 end
 
+"""
```
Comment (nalimilan) [inline src/macros.jl:353]: I guess you'll also add docs to the manual?

File: src/macros.jl [hunk 11]
```diff
@@ -350,6 +350,99 @@ macro passmissing(args...)
...
+    astable(args...)
+
+Return a `NamedTuple` from a transformation inside DataFramesMeta.jl macros.
+
+`@astable` acts on a single block. It works through all top-level expressions
+and collects all such expressions of the form `:y = x`, i.e. assignments to a
+`Symbol`, which is a syntax error outside of the macro. At the end of the
```
Comment (nalimilan) [inline src/macros.jl:360]: Outside of DataFramesMeta macros, right?

**Review by nalimilan — COMMENTED** [review]
Looks pretty cool! This will require a breaking release, right?

Using `@astable` to ensure operations are run sequentially is clever. The name is a bit surprising for this, but well... I also hope the compilation overhead isn't too large.

**Comment by pdeffebach** [comment]
> Using @astable to ensure operations are run sequentially is clever. The name is a bit surprising for this, but well... I also hope the compilation overhead isn't too large.

This macro does more than that. It allows for local variables to per persistent as well. If I were to just force sequential transform calls, I would just create a `transform` call for each `:y = f(:x)` expression.

File: src/macros.jl (hunk 1 above)
Comment (pdeffebach) [reply src/macros.jl:432]: great. changed.

File: src/parsing_astable.jl (hunk 2 above)
Comment (pdeffebach) [reply src/parsing_astable.jl:1]: added.

File: test/astable_flag.jl (hunk 6 above)
Comment (pdeffebach) [reply test/astable_flag.jl:15]: I wanted to make sure it could handle an expression that wasn't just assignments.

File: src/parsing_astable.jl (hunk 3 above)
Comment (pdeffebach) [reply src/parsing_astable.jl]: Ah this isn't quite right. Notice the two different dicts used, `lhs_assignments` and `inputs_to_function`.

**Comment by pdeffebach** [comment]
This is ready for another round of reviews. I have added docs to the manual as well.

File: docs/src/index.md [hunk 12]
```diff
@@ -396,11 +397,37 @@ julia> @rtransform df @passmissing x = parse(Int, :x_str)
...
 ```
 
+## Creating multiple columns at once with `@astable`
+
+Often new variables may depend on the same intermediate calculations. `@astable` makes it easy to create multiple
+new variables in the same operation, yet have them share
+information. 
+
+In a single block, all assignments of the form `:y = f(:x)` 
+or `$y = f(:x)` at the top-level are generate new columns.
```
Comment (bkamins) [inline docs/src/index.md:407]:
```suggestion
or `$y = f(:x)` at the top-level generate new columns.
```

Comment (bkamins) [reply docs/src/index.md:407]: maybe add what `$y` has to resolve to (I understand it has to be `Symbol`, or strings are also accepted?)

File: src/macros.jl [hunk 13]
```diff
@@ -350,6 +350,120 @@ macro passmissing(args...)
     throw(ArgumentError("@passmissing only works inside DataFramesMeta macros."))
 end
 
+"""
+    astable(args...)
```
Comment (bkamins) [inline src/macros.jl:354]:
```suggestion
    @astable(args...)
```

File: src/macros.jl [hunk 14]
```diff
@@ -350,6 +350,120 @@ macro passmissing(args...)
...
+
+```
+@transform df @astable begin
+    m = mean(:x)
+    :x_demeaned = :x .- m
+    :x2_demeaned = :x2 .- m
+end
+```
+
+The creation of `:x_demeaned` and `:x2_demeaned` both share the variable `m`,
```
Comment (bkamins) [inline src/macros.jl:404]: maybe add that the original data frame has to have `:x` and `:x2` columns (to make things clear).

File: src/macros.jl [hunk 15]
```diff
@@ -350,6 +350,120 @@ macro passmissing(args...)
...
+transformations. `@astable` solves this problem
+
+@rtransform df @astable begin
+    :new_col_1 = :x + :y
+    :new_col_2 = :new_col_1 + :z
+end
+
+Column assignment in `@astable` follows the same rules as
+column assignment more generally. Construct a new column
+from a string by escaping it with `$DOLLAR`.
```
Comment (bkamins) [inline src/macros.jl:427]: can you add an example of this?

File: src/macros.jl [hunk 16]
```diff
@@ -391,7 +392,7 @@ transform(df, [:a] => ByRow(f) => AsTable)
 `@astable` has two major advantages at the cost of increasing complexity.
 First, `@astable` makes it easy to create multiple columns from a single
 transformation, which share a scope. For example, `@astable` allows
-for the following
+for the following (where `:x` and `:x_2` exist in the `DataFrame` already).
```
Comment (bkamins) [inline src/macros.jl:395]:
```suggestion
for the following (where `:x` and `:x_2` exist in the data frame already).
```

File: docs/src/index.md [hunk 17]
```diff
@@ -396,11 +397,38 @@ julia> @rtransform df @passmissing x = parse(Int, :x_str)
...
+   3 │     3    600      200       0
+```
+
+
 ## [Working with column names programmatically with `$`](@id dollar)
 
 DataFramesMeta provides the special syntax `$` for referring to 
-columns in a data frame via a `Symbol`, string, or column position as either
-a literal or a variable. 
+columns in a data frame via a `Symbol`, string, or column position as either a literal or a variable.
```
Comment (bkamins) [inline docs/src/index.md:431]: While we are at it given our recent discussion on Discourse, I think it is essential to mention when the `$` reference is resolved.
Also maybe add an example when macros are used within a function? I think these are cases not trivial. This can be another PR of course

File: src/macros.jl [hunk 18]
```diff
@@ -350,6 +350,138 @@ macro passmissing(args...)
...
+transformations. `@astable` solves this problem
+
+@rtransform df @astable begin
+    :new_col_1 = :x + :y
+    :new_col_2 = :new_col_1 + :z
+end
+
+Column assignment in `@astable` follows the same rules as
+column assignment more generally. Construct a new column
+from a string by escaping it with `$DOLLAR`, which can be a
```
Comment (bkamins) [inline src/macros.jl:428]: "from a string" here is unclear.

File: src/macros.jl [hunk 19]
```diff
@@ -350,6 +350,138 @@ macro passmissing(args...)
...
 end
 
+"""
+    @astable(args...)
+
+Return a `NamedTuple` from a single transformation inside the DataFramesMeta.jl
+macros, `@select`, `@transform`, and their mutating and row-wise equivalents.
+
+`@astable` acts on a single block. It works through all top-level expressions
+and collects all such expressions of the form `:y = ...`, i.e. assignments to a
```
Comment (bkamins) [inline src/macros.jl:360]: also mention the `$` form on LHS.

File: src/macros.jl [hunk 20]
```diff
@@ -350,6 +350,138 @@ macro passmissing(args...)
...
+2×3 DataFrame
+ Row │ a      min_b  max_b
+     │ Int64  Int64  Int64
+─────┼─────────────────────
+   1 │     1      5      6
+   2 │     2     70     80
+
+julia> @rtransform df @astable begin
+           f_a = first(:a)
+           $(DOLLAR)new_col = :a + :b + f_a
```
Comment (bkamins) [inline src/macros.jl:467]: where is `new_col` defined?

File: docs/src/index.md (hunk 17 above)
Comment (pdeffebach) [reply docs/src/index.md:431]: I will do this as another PR. In summary, you can't use other macros which use `$`. I will try and sort out if I can carve out an exception.

File: src/macros.jl (hunk 19 above)
Comment (pdeffebach) [reply src/macros.jl:360]: added.

File: src/macros.jl (hunk 18 above)
Comment (pdeffebach) [reply src/macros.jl:428]: I have re-phrased. Though I think people reading this are probably familiar with column assignment so I don't want to go into it in too much detail here.

File: src/macros.jl (hunk 20 above)
Comment (pdeffebach) [reply src/macros.jl:467]: Sorry. Fixed.

**Comment by pdeffebach** [comment]
Thanks, I've responded to the most recent round of reviews. 

I have added more tests. I can't think of any new tests to add at the moment, they seem pretty well covered.

**Comment by pdeffebach** [comment]
cc @jkrumbiegel, if you want to review.

File: src/macros.jl [hunk 21]
```diff
@@ -284,7 +284,7 @@ end
 
 
 """
-    passmissing(args...)
+    @passmissing(args...)
 
 Propograte missing values inside DataFramesMeta.jl macros.
```
Comment (nalimilan) [inline src/macros.jl:289]:
```suggestion
Propagate missing values inside DataFramesMeta.jl macros.
```

File: src/macros.jl (hunk 20 above)
Comment (nalimilan) [reply src/macros.jl:467]: I still don't see it.

File: src/parsing_astable.jl [hunk 22]
```diff
@@ -43,9 +50,15 @@ function is_column_assigment(ex::Expr)
...
 # Taken from MacroTools.jl
-# No docstring so assumed untable
+# No docstring so assumed unstable
 block(ex) = isexpr(ex, :block) ? ex : :($ex;)
 
+sym_or_str_to_sym(x::Union{AbstractString, Symbol}) = Symbol(x)
+function sym_or_str_to_sym(x)
+    e = "New columns created inside @astable must be Symbols or AbstractStrings"
+    throw(ArgumentError(e))
+end
```
Comment (nalimilan) [inline src/parsing_astable.jl:60]:
```suggestion
sym_or_str_to_sym(x) =
    throw(ArgumentError("New columns created inside @astable must be Symbols or AbstractStrings"))
```

File: test/astable_flag.jl [hunk 23]
```diff
@@ -120,5 +120,50 @@ end
...
+
+    gd = groupby(df, :a)
+
+    d = @combine gd @astable begin
+        ex = extrema(:b)
+        :b_min = ex[1]
+        :b_max = ex[2]
+    end
+
+    @test sort(d.b_min) == [5, 7]
```
Comment (nalimilan) [inline test/astable_flag.jl:134]: This kind of test is quite fragile. Better sort the whole data frame and compare it to the reference to make sure that groups and values match.

**Comment by nalimilan** [comment]
> This macro does more than that. It allows for local variables to per persistent as well. If I were to just force sequential transform calls, I would just create a `transform` call for each `:y = f(:x)` expression.

Yes I know. What I'm saying is that mentioning sequential operations as the main justification for it was a bit weird.

**Comment by pdeffebach** [comment]
I don't understand the changes to th egithub review interface, but I think i've addressed lingering issues. 

I just added some checks to make sure you can't do `@passmissing` and `@astable` at the same time. I think this is intuitive in the `@byrow` case, but I want to finish #276 , which makes `@passmissing` work on column-wise transformations, before I do this.

**Comment by pdeffebach** [comment]
Tests pass and this can be merged.

File: docs/src/index.md (hunk 17 above)
Comment (bkamins) [reply docs/src/index.md:431]: To be clear why I stress it so much. With DataFrames.jl my answer to users is: if you learn Julia Base then you will know exactly how DataFrames.jl works. With DataFramesMeta.jl unfortunately this is not the case as it is a DSL so we need to be very precise how things work in documentation.

File: src/macros.jl (hunk 20 above)
Comment (bkamins) [reply src/macros.jl:484]: I see it. It is fixed :)

File: test/astable_flag.jl [hunk 24]
```diff
@@ -0,0 +1,176 @@
...
+        :x = [5]
+        y = 100
+        nothing
+    end
+
+    @test d == DataFrame(x = 5)
+end
+
+@testset "@astable with just assignments, mutating" begin
+    # After finalizing above testset
```
Comment (bkamins) [inline test/astable_flag.jl:46]: @pdeffebach - this seems to be WIP?

File: test/astable_flag.jl [hunk 25]
```diff
@@ -0,0 +1,178 @@
...
+using Statistics
+
+const ≅ = isequal
+
+@testset "@astable with just assignments" begin
+    df = DataFrame(a = 1, b = 2)
+
+    d = @rtransform df @astable begin
+        :x = 1
+
```
Comment (bkamins) [inline test/astable_flag.jl:14]: newline is not needed. Also I am not clear why you add `nothing` here and below? Does it change anything?

File: test/astable_flag.jl [hunk 26]
```diff
@@ -0,0 +1,178 @@
...
+        :x = 1
+        y = 5
+        :z = :x + y
+    end
+
+    @test d == DataFrame(x = 1, z = 6)
+
+    x_str = "x"
+    d = @rtransform df @astable begin
+        $x_str = 1
```
Comment (bkamins) [inline test/astable_flag.jl:107]: this test is not good as you are assigning `1` to `:x` column, and it already was `1` so we do not see if it works correctly. Same below.

**Review by pdeffebach — COMMENTED** [review]
Thanks for the round of reviews. I think I have addressed everything. Sorry about forgetting some of the tests.

**Review by bkamins — APPROVED** [review]
Looks good. Just please check why Documenter fails before merging. Thank you!

File: src/macros.jl [hunk 27]
```diff
@@ -282,11 +282,10 @@ macro byrow(args...)
     throw(ArgumentError("@byrow is deprecated outside of DataFramesMeta macros."))
 end
 
-
 """
     @passmissing(args...)
 
-Propograte missing values inside DataFramesMeta.jl macros.
+Propagrate missing values inside DataFramesMeta.jl macros.
```
Comment (nalimilan) [inline src/macros.jl:288]:
```suggestion
Propagate missing values inside DataFramesMeta.jl macros.
```

File: src/macros.jl [hunk 28]
```diff
@@ -350,18 +349,24 @@ macro passmissing(args...)
     throw(ArgumentError("@passmissing only works inside DataFramesMeta macros."))
 end
 
+global astable_docstring_snippet = """
```
Comment (nalimilan) [inline src/macros.jl:352]: Did you mean this?
```suggestion
const astable_docstring_snippet = """
```

File: src/macros.jl [hunk 29]
```diff
@@ -350,18 +349,24 @@ macro passmissing(args...)
...
+    Transformations can also use the macro-flag `@astable` for creating multiple
```
Comment (nalimilan) [inline src/macros.jl:353]:
```suggestion
    Transformations can also use the macro-flag [`@astable`](@ref) for creating multiple
```

File: src/parsing.jl [hunk 30]
```diff
@@ -224,6 +226,8 @@ function get_source_fun(function_expr; exprflags = deepcopy(DEFAULT_FLAGS))
         end
     end
 
+    println(MacroTools.prettify(fun))
+
```
Comment (nalimilan) [inline src/parsing.jl:230]:
```suggestion
```

**Comment by pdeffebach** [comment]
Thanks for the review. @nalimilan ready to merge?

**Review by nalimilan — APPROVED** [review]

**Comment by bkamins** [comment]
Bravo!
